apis/python/tests/test_shape.py

```diff
@@ -0,0 +1,205 @@
+
+from tests._util import maybe_raises
+
+
+@pytest.mark.parametrize(
+    "element_dtype",
+    [
+        pa.float64(),
+        pa.float32(),
+        pa.int64(),
+        pa.uint16(),
+    ],
+)
+@pytest.mark.parametrize(
+    "shape_exc",
+    [
+        # Note: non-None exceptions are coming on https://github.com/single-cell-data/TileDB-SOMA/issues/2407
+        [(100,), None],
+        [(100, 200), None],
+        [(100, 200, 300), None],
+    ],
+)
+def test_sparse_nd_array_basics(
+    tmp_path,
+    element_dtype,
+    shape_exc,
+):
+    uri = tmp_path.as_posix()
+    arg_shape, arg_create_exc = shape_exc
+    ndim = len(arg_shape)
+
+    # Create the array
+    with maybe_raises(arg_create_exc):
+        snda = tiledbsoma.SparseNDArray.create(
+            uri,
+            type=element_dtype,
+            shape=arg_shape,
+        )
+    if arg_create_exc is not None:
+        return
+
+    assert tiledbsoma.SparseNDArray.exists(uri)
+
+    # Test the various accessors
+    with tiledbsoma.SparseNDArray.open(uri) as snda:
+        assert snda.shape == arg_shape
+
+        # Before current-domain support: shape is maxshape.
+        #
+        # With current-domain support:
+        # We expect XXX to be set to a big signed int32. (There are details on the exact value of
+        # that number, involving R compatibility, and leaving room for a single tile capacity, etc
+        # ...  we could check for some magic value but it suffices to check that it's over 2
+        # billion.)
+        assert snda.shape == snda.maxshape
+        #for e in snda.maxshape:
+        #    assert e > 2_000_000_000
+
+        # No data have been written for this test case
+        assert snda.non_empty_domain() == tuple([(0, 0)] * ndim)
+
+    # soma_dim_0: (0,1)
+    # soma_dim_1: (2,3)
+    # soma_dim_2: (4,5)
+    coords = []
+    dim_names = []
+    for i in range(ndim):
+        dim_names.append(f"soma_dim_{i}")
+        coords.append((2 * i, 2 * i + 1))
+    coords = tuple(coords)
+
+    # Write some data
+    with tiledbsoma.SparseNDArray.open(uri, "w") as snda:
+        dikt = {"soma_data": [4, 5]}
+        for i in range(ndim):
+            dikt[dim_names[i]] = coords[i]
+        table = pa.Table.from_pydict(dikt)
+        snda.write(table)
+
+    # Test the various accessors
+    with tiledbsoma.SparseNDArray.open(uri) as snda:
+        assert snda.shape == arg_shape
+        # This will change with current-domain support
+        assert snda.shape == snda.maxshape
+        #for e in snda.maxshape:
+        #    assert e > 2_000_000_000
+        assert snda.non_empty_domain() == coords
+
+    # Test reads out of bounds
+    with tiledbsoma.SparseNDArray.open(uri) as snda:
+        # TODO: make sure this doesn't come through as RuntimeError
+        # https://github.com/single-cell-data/TileDB-SOMA/issues/2407
+        with pytest.raises((RuntimeError, ValueError)):
+            coords = tuple([arg_shape[i] + 10 for i in range(ndim)])
```

Comment:
nit: don't need an intermediary list for tuple generators

```suggestion
            coords = tuple(arg_shape[i] + 10 for i in range(ndim))
```